src/GatewayAPI/Services/CoreNodesSupervisorService.cs

```diff
@@ -0,0 +1,209 @@
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using Common.Extensions;
+using GatewayAPI.Configuration;
+using GatewayAPI.Configuration.Models;
+using GatewayAPI.CoreCommunications;
+using CoreApiModel = RadixCoreApi.Generated.Model;
+
+namespace GatewayAPI.Services;
+
+public interface ICoreNodesSupervisorService
+{
+    public CoreApiNode GetRandomHealthiestCoreNode();
+
+    public Task ReviseCoreNodesHealth(CancellationToken cancellationToken);
+}
+
+public class CoreNodesSupervisorService : ICoreNodesSupervisorService
+{
+    // Using explicit integers for enum values
+    // because they're used for ordering the nodes (from best to worst).
+    private enum CoreNodeStatus
+    {
+        HealthyAndSynced = 0,
+        Unknown = 1,
+        HealthyButLagging = 2,
+        Unhealthy = 3,
+    }
+
+    private static readonly IEnumerable<CoreNodeStatus> _statusesFromBestToWorst =
+        Enum.GetValues(typeof(CoreNodeStatus)).Cast<CoreNodeStatus>().OrderBy(s => s);
+
+    private readonly ILogger _logger;
+    private readonly HttpClient _httpClient;
+    private readonly CoreApiModel.NetworkIdentifier _networkIdentifier;
+    private readonly long _maxAllowedStateVersionLagToBeConsideredSynced;
+
+    private readonly Dictionary<CoreApiNode, CoreNodeStatus> _coreNodesStatuses;
+    private readonly object _coreNodesStatusesLock = new();
+
+    public CoreNodesSupervisorService(
+        ILogger<CoreNodesSupervisorService> logger,
+        HttpClient httpClient,
+        IGatewayApiConfiguration configuration)
+    {
+        _logger = logger;
+        _httpClient = httpClient;
+
+        _networkIdentifier = new CoreApiModel.NetworkIdentifier(configuration.GetNetworkName());
+        _maxAllowedStateVersionLagToBeConsideredSynced = configuration
+            .GetCoreApiNodeHealth().MaxAllowedStateVersionLagToBeConsideredSynced;
+
+        _coreNodesStatuses = configuration.GetCoreNodes()
+            .Where(n => n.IsEnabled && !string.IsNullOrWhiteSpace(n.CoreApiAddress))
+            .ToDictionary(n => n, _ => CoreNodeStatus.Unknown);
+    }
+
+    public CoreApiNode GetRandomHealthiestCoreNode()
+    {
+        lock (_coreNodesStatusesLock)
+        {
+            foreach (var status in _statusesFromBestToWorst)
```

Comment:
We possibly want a config option to exclude unhealthy and/or lagging from being used at all - and modify the exception message appropriately.

By default, I'd guess we should only allow requests to unknown or synced up nodes 👍